### PR TITLE
Cuda time measurements

### PR DESCRIPTION
**Pull request by AliinCern**
#### What is the purpose of the program?

A device (GPU) in a local machine can read/write host (CPU) memory in different methods by using Cuda API. In this program, we measured these methods in a local machine in order to find the best approachable to memory with respect to time. The program has four directions in time measurements:

1. From Host To Device.
2. On the Device.
3. On the Host.
4. From Device to Host.

#### What methods are used in the Program?

> With respect of First Direction Measurement:
Part 0) The Device reads from Host memory directly without using Cuda API.
Part 1) Measuring time to copy data from Host memory to Device memory using (cudaMemcpy).
Part 2) Measuring time to lock Host pages for Device to read/write data using (cudaHostRegister).
Part 3) Measuring time to copy from Host memory to Host page-lock using (cudaMemcpy).
Part 4) Measuring time to copy from Host memory to Host page-lock using (memcpy), then Measuring time to copy from Host 
page-lock to Device memory using (cudaMemcpy).
Part 5) Measuring time to lock Host pages using (cudaHostRegister) and Measuring time to copy from Host page-lock to Device memory using (cudaMemcpy). 

> With respect of Second Direction Measurement:
All Parts) Measuring time operations that are done on the Device using (cudaEventRecord).

> With respect of Third Direction Measurement:
All Parts) Measuring time operations that are done on the Device using Host time elapse.

> With respect of Fourth Direction Measurement:
Part 0) The Device writes on Host memory directly without using Cuda API.
Part 1) Measuring time to copy data from Device memory to Host memory using (cudaMemcpy).
Part 2) Measuring time to unlock Host pages using (cudaHostUnregister).
Part 3) Measuring time to copy from Host page-lock to Host memory using (cudaMemcpy).
Part 4) Measuring time to copy from Device memory to Host page-lock using (cudaMemcpy), then Measuring time to copy from Host page-lock to Host memory using (memcpy).
Part 5) Measuring time to copy from Device memory to Host page-lock using (cudaMemcpy), then unlock Host pages using (cudaHostUnregister).

#### What are the command line Options to run the Program?
  -p PARTS     select which parts (methods) to run, for example -p 12345 will run all methods. Please note that to run Part 0, use 
                      number 6 for example -p 6
  -a COUNT   repeat each set of tasks COUNT times, and compute the average.
  -f                 to save the result to a file.
 
  -q                print the standard deviation of the measurements.
  -t TASKS      repeat the task on the device/gpu TASKS times.

#### Program Validation:
we ran the program using **_scram b runtests_**, and the result is this:
---> test cudaTimeMeasurement succeeded
TestTime:1
^^^^ End Test cudaTimeMeasurement ^^^^

### CODE REVIEW

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
@cmsbuild, please test

File: HeterogeneousCore/CUDACore/test/cudaTimeMeasurement.cu [hunk 1]
```diff
@@ -0,0 +1,1005 @@
...
+
+void printoutAll(std::vector<Timing> &timing, bool standerDeviationPrint);
+
+int getNumberofDigits(double number);
+
+void newLineTitle(int line, const std::string &title);
+
+void printResultEach(std::vector<Timing> &timing, int type, bool standerDeviationPrint);
+
+int main(int argc, char *argv[]) {
```
Comment (fwyzard) [inline HeterogeneousCore/CUDACore/test/cudaTimeMeasurement.cu:113]:
```suggestion
int main(int argc, char *argv[]) {
  cms::cudatest::requireDevices();
  
```

**Review by fwyzard — CHANGES_REQUESTED** [review]
The new test can only work if the machine has a GPU...
@AliinCern could you add `cms::cudatest::requireDevices();` at the beginning of the `main()` functions ?

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
@cmsbuild, please test

File: HeterogeneousCore/CUDACore/test/BuildFile.xml [hunk 2]
```diff
@@ -15,4 +15,11 @@
     <use name="cuda"/>
   </bin>
 
+  <bin name="cudaTimeMeasurement" file="cudaTimeMeasurement.cu">
+    <use name="cuda"/>
```
Comment (fwyzard) [inline HeterogeneousCore/CUDACore/test/BuildFile.xml:19]: My bad, I forgot that you need to add here:
```xml
    <use name="HeterogeneousCore/CUDAUtilities"/>
```

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by fwyzard** [comment]
@smuzaffar is there a way to request running the unit tests on the GPU machine ?

**Comment by smuzaffar** [comment]
> @smuzaffar is there a way to request running the unit tests on the GPU machine ?

No @fwyzard , currently there is no way to requests for unit tests to be run on GPU. As for PR tests, we are not deploying the `cmssw/tmp` directory under `/cvmfs/cms-ci.cern.ch` which means one needs to re-build on GPU machine to run tests. It is on my todo list to add support for unit tests for GPUs but will do when find some time

**Comment by perrotta** [comment]
+1
- It adds measurement tools in the test area
- Verified working correctly by the author in a private test
- Unit test not actually tested in bot because those unit tests do not run on GPU machines

**Pull request merged**